File: R/rdodo/tests/testhat/test-create_aircraft.R
require(testthat)
context("create_aircraft function")

test_that("the create_aircraft function works", {

  # TODO: check for Bluebird at the expected url and stop if unavailable.

  aircraft_id <- "test1234"
  type <- "B744"
  latitude <- 0
  longitude <- 0
  heading <- 0
  flight_level <- 250
  speed <- 200

  expect_true(create_aircraft(aircraft_id = aircraft_id,
                              type = type,
                              latitude = latitude,
                              longitude = longitude,
                              heading = heading,
                              flight_level = flight_level,
                              speed = speed))

})

test_that("the create_aircraft latitude guard clause works", {

  # TODO: check for Bluebird at the expected url and stop if unavailable.

  aircraft_id <- "test1234"
  type <- "B744"
  longitude <- 0
  heading <- 0
  flight_level <- 250
  speed <- 200

  latitude <- 91
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                              type = type,
                              latitude = latitude,
                              longitude = longitude,
                              heading = heading,
                              flight_level = flight_level,
                              speed = speed))

  latitude <- -90.1
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               flight_level = flight_level,
                               speed = speed))

  # Valid latitude:
  latitude <- -90
  expect_true(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               flight_level = flight_level,
                               speed = speed))
})

test_that("the create_aircraft longitude guard clause works", {

  # TODO: check for Bluebird at the expected url and stop if unavailable.

  aircraft_id <- "test1234"
  type <- "B744"
  latitude <- 0
  heading <- 0
  flight_level <- 250
  speed <- 200

  longitude <- 180
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               flight_level = flight_level,
                               speed = speed))

  longitude <- -180.1
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               flight_level = flight_level,
                               speed = speed))

  # Valid longitude:
  longitude <- -180
  expect_true(create_aircraft(aircraft_id = aircraft_id,
                              type = type,
                              latitude = latitude,
                              longitude = longitude,
                              heading = heading,
                              flight_level = flight_level,
                              speed = speed))
})

test_that("the create_aircraft heading guard clause works", {

  # TODO: check for Bluebird at the expected url and stop if unavailable.

  aircraft_id <- "test1234"
  type <- "B744"
  latitude <- 0
  longitude <- 0
  flight_level <- 250
  speed <- 200

  heading <- -0.1
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               flight_level = flight_level,
                               speed = speed))

  heading <- 360
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               flight_level = flight_level,
                               speed = speed))

  # Valid heading:
  heading <- 359.99
  expect_true(create_aircraft(aircraft_id = aircraft_id,
                              type = type,
                              latitude = latitude,
                              longitude = longitude,
                              heading = heading,
                              flight_level = flight_level,
                              speed = speed))
})

test_that("the create_aircraft altitude guard clause works", {

  # TODO: check for Bluebird at the expected url and stop if unavailable.

  aircraft_id <- "test1234"
  type <- "B744"
  latitude <- 0
  longitude <- 0
  heading <- 0
  speed <- 200

  altitude <- 6000
  flight_level <- 250
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               altitude = altitude,
                               flight_level = flight_level,
                               speed = speed))

  altitude <- 6001
  flight_level <- NULL
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               altitude = altitude,
                               flight_level = flight_level,
                               speed = speed))

  altitude <- -1
  flight_level <- NULL
  expect_error(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               altitude = altitude,
                               flight_level = flight_level,
                               speed = speed))

  # Valid altitude:
  altitude <- 0
  flight_level <- NULL
  expect_true(create_aircraft(aircraft_id = aircraft_id,
                              type = type,
                              latitude = latitude,
                              longitude = longitude,
                              heading = heading,
                              altitude = altitude,
                              flight_level = flight_level,
                              speed = speed))

  # Valid altitude:
  altitude <- 6000
  flight_level <- NULL
  expect_true(create_aircraft(aircraft_id = aircraft_id,
                               type = type,
                               latitude = latitude,
                               longitude = longitude,
                               heading = heading,
                               altitude = altitude,
                               flight_level = flight_level,
                               speed = speed))
})

File: PyDodo/tests/unit/test_create_aircraft.py
"""
Test create_aircraft function raises error if incorrect inputs are provided.

Function returns True if valid inputs are used and connection to bluebird exists.
"""

import pytest
import numpy as np

from pydodo.create_aircraft import create_aircraft
from pydodo.utils import ping_bluebird

# Valid input parameter values
aircraft_id = "TST1001"
type = "B744"
latitude = 0
longitude = 0
heading = 0
altitude = np.nan
flight_level = 250
speed = 200

bb_resp = ping_bluebird()
@pytest.mark.skipif(not bb_resp, reason="Can't connect to bluebird")
def test_output_create_aircraft():
    """
    For valid inputs, check function returns True (if can connect to bluebird)
    """
    output = create_aircraft(aircraft_id, type, latitude, longitude,
                             heading, altitude, flight_level, speed)
    assert output == True

@pytest.mark.parametrize(
    "acid,tp",
    [("", "B744"), ("TST1001", "")]
)
def test_input_aircraft_info(acid, tp):
    """
    Check incorrect aircraft_id and type inputs raise error
    """
    with pytest.raises(AssertionError):
        create_aircraft(acid, tp, latitude, longitude,
                        heading, altitude, flight_level, speed)

@pytest.mark.parametrize("lat", [91, -90.1])
def test_input_latitude(lat):
    """
    Check incorrect latitude inputs raise error
    """
    with pytest.raises(AssertionError):
        create_aircraft(aircraft_id, type, lat, longitude,
                        heading, altitude, flight_level, speed)

@pytest.mark.parametrize("lon", [180, -180.1])
def test_input_longitude(lon):
    """
    Check incorrect longitude inputs raise error
    """
    with pytest.raises(AssertionError):
        create_aircraft(aircraft_id, type, latitude, lon,
                        heading, altitude, flight_level, speed)

@pytest.mark.parametrize("hdg", [-0.1, 360])
def test_input_heading(hdg):
    """
    Check incorrect heading inputs raise error
    """
    with pytest.raises(AssertionError):
        create_aircraft(aircraft_id, type, latitude, longitude,
                        hdg, altitude, flight_level, speed)

@pytest.mark.parametrize(
    "alt,fl",
    [(6000,250), (6001, np.nan), (np.nan, 59)]
    )
def test_input_alt_fl(alt, fl):
    """
    Check incorrect combination of altitude and flight level inputs raise error
    """
    with pytest.raises(AssertionError):
        create_aircraft(aircraft_id, type, latitude, longitude,
                        heading, alt, fl, speed)

@pytest.mark.parametrize("spd", [-0.1])
def test_input_speed(spd):
    """
    Check incorrect speed inputs raise error
    """
    with pytest.raises(AssertionError):
        create_aircraft(aircraft_id, type, latitude, longitude,
                        heading, altitude, flight_level, spd)
